### PR TITLE
[Inductor] Align floating-point sum reduction signed-zero semantics with eager (#187721)

### LINKED ISSUE #187721: [Inductor] Silent correctness : F.kl_div reduction over -0.0 yields different signbits in Eager vs Compiled
### 🐛 Describe the bug

When using `F.kl_div` with `reduction='mean'` and `log_target=True` in `float16`, the output in Eager mode evaluates to `0.0` (signbit: False), whereas the `torch.compile` (Inductor) evaluates to `-0.0` (signbit: True).

Although `0.0` and `-0.0` evaluate to `True` under `torch.allclose`, this signbit difference causes severe discrepancies in downstream operations that are sensitive to the sign of zero (such as `torch.copysign`, `torch.atan2`). For example, applying `torch.copysign` to the output yields `1.0` in Eager mode but `-1.0` in Compiled mode.

*Note: This issue involves a signbit discrepancy like #187336, but the manifestation is opposite (Eager is `0.0`, Compiled is `-0.0`) and it occurs during a reduction operation rather than a pointwise operation.*

**Reproduce script:**
```python
import torch
import torch.nn.functional as F
import warnings

warnings.filterwarnings("ignore")

print("torch version:", torch.__version__)
if torch.cuda.is_available():
    print("cuda version:", torch.version.cuda)

def fn(input_t, target_t):
    y = F.kl_div(input_t, target_t, reduction='mean', log_target=True)
    sign_val = torch.copysign(torch.tensor(1.0, device=input_t.device, dtype=torch.float32), y)
    return sign_val, y

device = "cuda" if torch.cuda.is_available() else "cpu"
dtype = torch.float16

input_t = torch.tensor([100.0, 100.0], dtype=dtype, device=device).requires_grad_(True)
target_t = torch.tensor([-100.0, -100.0], dtype=dtype, device=device)

y_eager_sign, y_eager = fn(input_t, target_t)

compiled_fn = torch.compile(fn, backend="inductor", fullgraph=True)
y_compiled_sign, y_compiled = compiled_fn(input_t, target_t)

print("=== Raw kl_div Output ===")
print(f"Eager:    {y_eager.item()} (signbit: {y_eager.signbit().item()})")
print(f"Compiled: {y_compiled.item()} (signbit: {y_compiled.signbit().item()})")

print("\n=== Downstream copysign Output ===")
print(f"Eager copysign:    {y_eager_sign.item()}")
print(f"Compiled copysign: {y_compiled_sign.item()}")

print("\n=== Match Verification ===")
print("torch.allclose passes: ", torch.allclose(y_eager, y_compiled))
print("Actual values match (Equal): ", torch.equal(y_eager_sign, y_compiled_sign))
```

**Output logs:**
```text
torch version: 2.14.0.dev20260617+cu130
cuda version: 13.0
=== Raw kl_div Output ===
Eager:    0.0 (signbit: False)
Compiled: -0.0 (signbit: True)

=== Downstream copysign Output ===
Eager copysign:    1.0
Compiled copysign: -1.0

=== Match Verification ===
torch.allclose passes:  True
Actual values match (Equal):  False
```

### Versions

PyTorch version: 2.14.0.dev20260617+cu130
CUDA used to build PyTorch: 13.0
OS: Ubuntu 24.04.4 LTS (x86_64)
Is CUDA available: True
CUDA runtime version: 12.0.140

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by patrickswedish**
### Summary

Fix signed-zero parity for floating-point Inductor sum reductions across both the small unrolled reduction path and Triton reduction codegen.

Fixes #187721.

### Root Cause

ATen/eager sum reduction semantics effectively include a `+0.0` additive identity.

Inductor had two paths that could instead preserve `-0.0`:

1. Small reductions lowered through `_unroll_reduction_fn` used `functools.reduce` without a `+0.0` initializer.
2. Triton final reduction could return the result of `tl.sum` directly, allowing an all-negative-zero reduction to remain `-0.0`.

That produced observable differences in sign-sensitive downstream operations such as `copysign` and `atan2`, including KL-div underflow cases.

### Fix

- Normalize floating-point unrolled SUM reductions with a typed `+0.0` additive identity in `_unroll_reduction_fn`.
- Normalize floating-point Triton SUM final reductions with `+0.0`.
- Preserve integer reductions and all unrelated reduction behavior.
- Use `dtype.is_floating_point` directly.
- Document the signed-zero normalization so the apparently redundant `+0.0` is not removed accidentally.

### `strict_signed_zero`

`config.strict_signed_zero` applies to comparison-based reductions such as min/max, where equal signed-zero values require tie-breaking behavior.

SUM is different: `+0.0` is the additive identity already reflected by eager/ATen sum semantics.

This patch therefore restores default eager parity for SUM rather than gating the fix behind `strict_signed_zero`.

### Regression Coverage

Parameterized coverage across:

- `float16`
- `bfloat16`
- `float32`
- `float64`

Cases include:

- small reductions below the unroll threshold
- large 1024-element Triton reductions
- dimension reductions
- KL-div signed-zero parity
- arithmetic corner cases including infinities, NaNs, and mixed signed zeros

The tests contain direct assertions for expected positive-zero sign semantics rather than only eager-vs-compiled equality.

### RED / GREEN Proof

Without the production fixes:

- small/unrolled SUM: eager `+0.0`, compiled `-0.0`
- large Triton SUM: eager `+0.0`, compiled `-0.0`

With the fix restored:

- **23 focused GPU regression cases pass on NVIDIA RTX A2000**.

### Validation

- `pytest test/inductor/test_cuda_repro.py -k test_kl_div_reduction_signbit_parity_187721` — passed
- `pytest test/inductor/test_cpu_repro.py -k test_kl_div_reduction_signbit_parity_187721` — passed
- `git diff --check` — clean

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by eellison — CHANGES_REQUESTED** [review]
sorry, not accepting prs for this.

**Comment by patrickswedish** [comment]
Thanks for the detailed review — the coverage issue was valid. The original 2-element tests were lowered through the unrolled pointwise path and did not exercise `TritonKernel.final_reduction`.

I’ve updated the patch to address both reduction paths:

- floating-point sums in `_unroll_reduction_fn` now absorb `-0.0` into the `+0.0` additive identity;
- the Triton reduction path retains the corresponding normalization, now using `dtype.is_floating_point` directly with a comment documenting the signed-zero invariant.

I also replaced the original coverage with parameterized regressions across `float16`, `bfloat16`, `float32`, and `float64`, covering small unrolled reductions, 1024-element Triton reductions, and `dim=` reductions with direct expected `signbit == False` assertions.

RED/GREEN was verified on an NVIDIA RTX A2000:

- without the fixes, both the small unrolled and large Triton cases return `-0.0` while eager returns `+0.0`;
- with the fixes restored, the focused battery passes: **23 passed**.

I also checked `strict_signed_zero`: that setting handles min/max tie-breaking semantics; sum is different because `+0.0` is the additive identity used by eager ATen, so this restores default sum parity without adding comparison/selection logic.

Thanks again for catching the unrolled-path gap.

**Comment by patrickswedish** [comment]
Thanks for the clarification. Understood — I’ll close this out. Appreciate you taking a look.
